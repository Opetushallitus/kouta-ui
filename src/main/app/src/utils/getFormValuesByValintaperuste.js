import { isArray, isObject, mapValues } from 'lodash';
import parseEditorState from './draft/parseEditorState';
import getValintakoeFieldsValues from './getValintakoeFieldsValues';

const parseSisalto = ({ sisalto }) => {
  if (!isArray(sisalto)) {
    return [];
  }

  return sisalto.map(({ tyyppi, data }) => {
    if (tyyppi === 'teksti') {
      return {
        tyyppi,
        data: isObject(data) ? mapValues(data, parseEditorState) : {},
      };
    }

    return { tyyppi, data };
  });
};

const getFormValuesByValintaperuste = valintaperuste => {
  const {
    hakutapaKoodiUri = null,
    kielivalinta = [],
    kohdejoukkoKoodiUri = null,
    nimi = {},
    metadata = {},
    koulutustyyppi = null,
    onkoJulkinen = false,
    tila,
    sorakuvausId,
    valintakokeet,
  } = valintaperuste;

  const { valintatavat = [], kuvaus = {} } = metadata;

  return {
    tila,
    perustiedot: {
      tyyppi: koulutustyyppi,
      kieliversiot: kielivalinta,
      hakutapa: hakutapaKoodiUri,
      kohdejoukko: kohdejoukkoKoodiUri ? { value: kohdejoukkoKoodiUri } : null,
    },
    julkinen: onkoJulkinen,
    kuvaus: {
      nimi,
      kuvaus: mapValues(kuvaus || {}, parseEditorState),
    },
    valintatavat: (valintatavat || []).map(
      ({
        kuvaus,
        nimi,
        valintatapaKoodiUri,
        sisalto,
        kynnysehto,
        enimmaispisteet,
        vahimmaispisteet,
      }) => ({
        kuvaus: kuvaus || {},
        nimi: nimi || {},
        kynnysehto: kynnysehto || {},
        tapa: valintatapaKoodiUri ? { value: valintatapaKoodiUri } : null,
        enimmaispistemaara: enimmaispisteet || '',
        vahimmaispistemaara: vahimmaispisteet || '',
        sisalto: parseSisalto({ sisalto }),
      }),
    ),
    soraKuvaus: sorakuvausId
      ? {
          value: sorakuvausId,
        }
      : null,
    valintakoe: getValintakoeFieldsValues(valintakokeet),
  };
};

export default getFormValuesByValintaperuste;
