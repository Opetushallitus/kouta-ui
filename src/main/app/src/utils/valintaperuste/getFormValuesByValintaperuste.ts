import _ from 'lodash';

import { parseEditorState } from '#/src/components/Editor/utils';
import { getKokeetTaiLisanaytotValues } from '#/src/utils/form/getKokeetTaiLisanaytotValues';

const parseSisalto = sisalto => {
  if (!_.isArray(sisalto)) {
    return [];
  }

  return sisalto.map(({ tyyppi, data }) => {
    if (tyyppi === 'teksti') {
      return {
        tyyppi,
        data: _.isObject(data) ? _.mapValues(data, parseEditorState) : {},
      };
    }

    return { tyyppi, data };
  });
};

export const getFormValuesByValintaperuste = valintaperuste => {
  const {
    hakutapaKoodiUri = null,
    kielivalinta = [],
    kohdejoukkoKoodiUri = null,
    nimi = {},
    metadata = {},
    koulutustyyppi = null,
    onkoJulkinen = false,
    tila,
    sorakuvausId,
    valintakokeet,
  } = valintaperuste;

  const {
    valintatavat = [],
    kuvaus = {},
    sisalto = [],
    valintakokeidenYleiskuvaus,
  } = metadata;

  return {
    tila,
    perustiedot: {
      tyyppi: koulutustyyppi,
      kieliversiot: kielivalinta,
      hakutapa: hakutapaKoodiUri,
      kohdejoukko: kohdejoukkoKoodiUri ? { value: kohdejoukkoKoodiUri } : null,
    },
    julkinen: onkoJulkinen,
    kuvaus: {
      nimi,
      kuvaus: _.mapValues(kuvaus || {}, parseEditorState),
      sisalto: parseSisalto(sisalto),
    },
    valintatavat: (valintatavat || []).map(
      ({
        nimi: valintatapaNimi,
        kuvaus: valintatapaKuvaus,
        sisalto: valintatapaSisalto,
        valintatapaKoodiUri,
        kynnysehto,
        enimmaispisteet,
        vahimmaispisteet,
      }) => ({
        nimi: valintatapaNimi || {},
        kuvaus: valintatapaKuvaus || {},
        sisalto: parseSisalto(valintatapaSisalto),
        tapa: valintatapaKoodiUri ? { value: valintatapaKoodiUri } : null,
        kynnysehto: _.mapValues(kynnysehto || {}, parseEditorState),
        enimmaispistemaara:
          _.toString(enimmaispisteet)?.replace('.', ',') || '',
        vahimmaispistemaara:
          _.toString(vahimmaispisteet)?.replace('.', ',') || '',
      })
    ),
    soraKuvaus: sorakuvausId
      ? {
          value: sorakuvausId,
        }
      : null,
    valintakokeet: getKokeetTaiLisanaytotValues(
      valintakokeet,
      valintakokeidenYleiskuvaus
    ),
  };
};
