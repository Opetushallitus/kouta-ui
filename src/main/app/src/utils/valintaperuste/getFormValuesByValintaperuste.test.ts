import getFormValuesByValintaperuste from '#/src/utils/valintaperuste/getFormValuesByValintaperuste';

test('getFormValuesByValintaperuste returns correct form values given valintaperuste', () => {
  const values = getFormValuesByValintaperuste({
    tila: 'tallennettu',
    hakutapaKoodiUri: 'tapa_1#1',
    kielivalinta: ['fi', 'sv'],
    kohdejoukkoKoodiUri: 'joukko_1#1',
    koulutustyyppi: 'tyyppi_1#1',
    onkoJulkinen: true,
    sorakuvausId: 'sora_1',
    metadata: {
      kielitaitovaatimukset: [
        {
          kieliKoodiUri: 'kieli_1#1',
          kielitaidonVoiOsoittaa: [
            {
              kielitaitoKoodiUri: 'osoitustapa_1#1',
              lisatieto: {},
            },
            {
              kielitaitoKoodiUri: 'osoitustapa_2#1',
              lisatieto: {},
            },
            {
              kielitaitoKoodiUri: 'kielitaidonosoittaminen_04#1',
              lisatieto: {
                fi: 'Fi kuvaus',
                sv: 'Sv kuvaus',
              },
            },
          ],
          vaatimukset: [
            {
              kielitaitovaatimusKoodiUri: 'tyyppi_1#1',
              kielitaitovaatimusKuvaukset: [
                {
                  kielitaitovaatimusKuvausKoodiUri: 'kuvaus_1#1',
                  kielitaitovaatimusTaso: 'erinomainen',
                },
              ],
            },
            {
              kielitaitovaatimusKoodiUri: 'tyyppi_2#1',
              kielitaitovaatimusKuvaukset: [
                {
                  kielitaitovaatimusKuvausKoodiUri: 'kuvaus_2#1',
                  kielitaitovaatimusTaso: 'ok',
                },
              ],
            },
          ],
        },
      ],
      kuvaus: {
        fi: '<h1>Fi kuvaus</h1>',
        sv: '<h1>Sv kuvaus</h1>',
      },
      osaamistaustaKoodiUrit: ['tausta_1#1', 'tausta_2#1'],
      valintatavat: [
        {
          enimmaispisteet: 20,
          kaytaMuuntotaulukkoa: false,
          kuvaus: {
            fi: 'Fi kuvaus',
            sv: 'Sv kuvaus',
          },
          kynnysehto: {
            fi: 'Fi kynnysehto',
            sv: 'Sv kynnysehto',
          },
          nimi: {
            fi: 'Fi nimi',
            sv: 'Sv nimi',
          },
          sisalto: [
            {
              data: {
                fi: '<h2>Fi sisalto</h2>',
                sv: '<h2>Sv sisalto</h2>',
              },
              tyyppi: 'teksti',
            },
            {
              data: {
                rows: [
                  {
                    columns: [
                      {
                        index: 0,
                        text: {
                          fi: 'Fi column1',
                          sv: 'Sv column1',
                        },
                      },
                      {
                        index: 1,
                        text: {
                          fi: 'Fi column2',
                          sv: 'Sv column2',
                        },
                      },
                    ],
                    index: 0,
                  },
                  {
                    columns: [
                      {
                        index: 0,
                        text: {
                          fi: 'Fi column3',
                          sv: 'Sv column3',
                        },
                      },
                      {
                        index: 1,
                        text: {
                          fi: 'Fi column4',
                          sv: 'Sv column4',
                        },
                      },
                    ],
                    index: 1,
                  },
                ],
              },
              tyyppi: 'taulukko',
            },
          ],
          vahimmaispisteet: 10,
          valintatapaKoodiUri: 'tapa_1#1',
        },
      ],
      valintakokeidenYleiskuvaus: {
        fi: '<p>Yleiskuvaus - fi</p>',
        sv: '<p>Yleiskuvaus - sv</p>',
      },
    },
    nimi: {
      fi: 'Fi nimi',
      sv: 'Sv nimi',
    },
    valintakokeet: [
      {
        nimi: {
          fi: 'valintakokeen nimi - fi',
          sv: 'valintakokeen nimi - sv',
        },
        tyyppiKoodiUri: 'tyyppi_1#1',
        metadata: {
          tietoja: {
            fi: '<p>Tietoa hakijalle - fi</p>',
            sv: '<p>Tietoa hakijalle - sv</p>',
          },
          liittyyEnnakkovalmistautumista: true,
          ohjeetEnnakkovalmistautumiseen: {
            fi: '<p>Ohjeet ennakkovalmistautumiseen - fi</p>',
            sv: '<p>ohjeet ennakkovalmistautumiseen - sv</p>',
          },
          erityisjarjestelytMahdollisia: true,
          ohjeetErityisjarjestelyihin: {
            fi: '<p>Ohjeet erityisjärjestelyihin - fi</p>',
            sv: '<p>Ohjeet erityisjärjestelyihin - sv</p>',
          },
        },
        tilaisuudet: [
          {
            osoite: {
              osoite: { fi: 'fi osoite', sv: 'sv osoite' },
              postinumeroKoodiUri: 'posti_1#1',
              postitoimipaikka: {
                fi: 'fi posititoimipaikka',
                sv: 'sv posititoimipaikka',
              },
            },
            aika: {
              alkaa: '2019-04-16T08:44',
              paattyy: '2019-04-18T08:44',
            },
            lisatietoja: {
              fi: 'fi lisatietoja',
              sv: 'sv lisatietoja',
            },
            jarjestamispaikka: {
              fi: 'jarjestamispaikka - fi',
              sv: 'jarjestamispaikka - sv',
            },
          },
        ],
      },
    ],
  });

  expect(values).toMatchSnapshot();
});
