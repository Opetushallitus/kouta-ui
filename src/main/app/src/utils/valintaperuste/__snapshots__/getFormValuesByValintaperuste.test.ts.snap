// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFormValuesByValintaperuste returns correct form values given valintaperuste 1`] = `
Object {
  "julkinen": true,
  "kuvaus": Object {
    "kuvaus": Object {
      "fi":         "<h1>Fi kuvaus</h1>",
      "sv":         "<h1>Sv kuvaus</h1>",
    },
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "sisalto": Array [],
  },
  "perustiedot": Object {
    "hakutapa": "tapa_1#1",
    "kieliversiot": Array [
      "fi",
      "sv",
    ],
    "kohdejoukko": Object {
      "value": "joukko_1#1",
    },
    "tyyppi": "tyyppi_1#1",
  },
  "soraKuvaus": Object {
    "value": "sora_1",
  },
  "tila": "tallennettu",
  "valintakokeet": Object {
    "kokeetTaiLisanaytot": Array [
      Object {
        "erityisjarjestelytMahdollisia": true,
        "id": undefined,
        "liittyyEnnakkovalmistautumista": true,
        "nimi": Object {
          "fi": "valintakokeen nimi - fi",
          "sv": "valintakokeen nimi - sv",
        },
        "ohjeetEnnakkovalmistautumiseen": Object {
          "fi":             "<p>Ohjeet ennakkovalmistautumiseen - fi</p>",
          "sv":             "<p>ohjeet ennakkovalmistautumiseen - sv</p>",
        },
        "ohjeetErityisjarjestelyihin": Object {
          "fi":             "<p>Ohjeet erityisjärjestelyihin - fi</p>",
          "sv":             "<p>Ohjeet erityisjärjestelyihin - sv</p>",
        },
        "tietoaHakijalle": Object {
          "fi":             "<p>Tietoa hakijalle - fi</p>",
          "sv":             "<p>Tietoa hakijalle - sv</p>",
        },
        "tilaisuudet": Array [
          Object {
            "alkaa": "2019-04-16T08:44",
            "jarjestamispaikka": Object {
              "fi": "jarjestamispaikka - fi",
              "sv": "jarjestamispaikka - sv",
            },
            "lisatietoja": Object {
              "fi":                 "<p>fi lisatietoja</p>",
              "sv":                 "<p>sv lisatietoja</p>",
            },
            "osoite": Object {
              "fi": "fi osoite",
              "sv": "sv osoite",
            },
            "paattyy": "2019-04-18T08:44",
            "postinumero": Object {
              "value": "posti_1#1",
            },
          },
        ],
        "tyyppi": Object {
          "value": "tyyppi_1#1",
        },
      },
    ],
    "yleisKuvaus": Object {
      "fi":         "<p>Yleiskuvaus - fi</p>",
      "sv":         "<p>Yleiskuvaus - sv</p>",
    },
  },
  "valintatavat": Array [
    Object {
      "enimmaispistemaara": 20,
      "kuvaus": Object {
        "fi": "Fi kuvaus",
        "sv": "Sv kuvaus",
      },
      "kynnysehto": Object {
        "fi":           "<p>Fi kynnysehto</p>",
        "sv":           "<p>Sv kynnysehto</p>",
      },
      "nimi": Object {
        "fi": "Fi nimi",
        "sv": "Sv nimi",
      },
      "sisalto": Array [
        Object {
          "data": Object {
            "fi":               "<h2>Fi sisalto</h2>",
            "sv":               "<h2>Sv sisalto</h2>",
          },
          "tyyppi": "teksti",
        },
        Object {
          "data": Object {
            "rows": Array [
              Object {
                "columns": Array [
                  Object {
                    "index": 0,
                    "text": Object {
                      "fi": "Fi column1",
                      "sv": "Sv column1",
                    },
                  },
                  Object {
                    "index": 1,
                    "text": Object {
                      "fi": "Fi column2",
                      "sv": "Sv column2",
                    },
                  },
                ],
                "index": 0,
              },
              Object {
                "columns": Array [
                  Object {
                    "index": 0,
                    "text": Object {
                      "fi": "Fi column3",
                      "sv": "Sv column3",
                    },
                  },
                  Object {
                    "index": 1,
                    "text": Object {
                      "fi": "Fi column4",
                      "sv": "Sv column4",
                    },
                  },
                ],
                "index": 1,
              },
            ],
          },
          "tyyppi": "taulukko",
        },
      ],
      "tapa": Object {
        "value": "tapa_1#1",
      },
      "vahimmaispistemaara": 10,
    },
  ],
}
`;
